File: Data_Structures/Heaps/reorganize_string.c
// Given a string S, check if the letters can be rearranged so that two characters that are adjacent
// to each other are not the same

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char ch;
    int count;
} CharCount;

int compare(const void* a, const void* b) {
    return ((CharCount*)b)->count - ((CharCount*)a)->count;
}

void swap(CharCount* a, CharCount* b) {
    CharCount temp = *a;
    *a = *b;
    *b = temp;
}

char* reorganizeString(char* S) {
    int charCounts[26] = {0};
    int len = strlen(S);

    for (int i = 0; i < len; i++)
        charCounts[S[i] - 'a']++;

    CharCount* counts = (CharCount*)malloc(26 * sizeof(CharCount));
    for (int i = 0; i < 26; i++) {
        counts[i].ch = 'a' + i;
        counts[i].count = charCounts[i];
    }

    qsort(counts, 26, sizeof(CharCount), compare);

    if (counts[0].count > (len + 1) / 2) {
        free(counts);
        return "";
    }

    int index = 0;
    for (int i = 0; i < 26; i++) {
        while (counts[i].count > 0) {
            S[index] = counts[i].ch;
            index += 2;
            if (index >= len)
                index = 1;
            counts[i].count--;
        }
    }

    free(counts);
    return S;
}

int main() {
    char S[] = "aab";

    char* result = reorganizeString(S);
    printf("Reorganized string: %s\n", result);

    return 0;
}

File: Data_Structures/Heaps/reorganize_string.py
# Given a string S, check if the letters can be rearranged so that two characters that are adjacent
# to each other are not the same

import heapq

def reorganize_string(S):
    char_counts = [0] * 26
    for ch in S:
        char_counts[ord(ch) - ord('a')] += 1

    heap = []
    for i in range(26):
        if char_counts[i] > 0:
            heapq.heappush(heap, (-char_counts[i], chr(ord('a') + i)))

    result = []
    while len(heap) >= 2:
        count1, ch1 = heapq.heappop(heap)
        count2, ch2 = heapq.heappop(heap)
        result.extend([ch1, ch2])
        if count1 + 1 < 0:
            heapq.heappush(heap, (count1 + 1, ch1))
        if count2 + 1 < 0:
            heapq.heappush(heap, (count2 + 1, ch2))

    if heap:
        count, ch = heapq.heappop(heap)
        if count < -1:
            return ""
        result.append(ch)

    return "".join(result)

S = "aab"
result = reorganize_string(S)
print("Reorganized string:", result)
